packages/nbt/src/checker/index.ts

```diff
@@ -286,18 +279,149 @@ export function blockStates(
 	}
 }
 
-/**
- * @param id If set to `undefined` or an empty array, all mcdoc compound definitions for this registry will be merged for checking, and unknown keys are allowed.
- */
+interface NbtPathLink {
+	path: NbtPathNode
+	node: NbtPathChild | { type: 'leaf'; range: core.Range }
+	prev?: NbtPathLink
+	next?: NbtPathLink
+}
+
+// TODO: check nbt index nodes and nbt compound nodes
 export function path(
 	registry: string,
 	id:
 		| core.FullResourceLocation
 		| readonly core.FullResourceLocation[]
 		| undefined,
-	options?: PathOptions,
 ): core.SyncChecker<NbtPathNode> {
 	return (node, ctx) => {
-		// TODO
+		// TODO: support dispatcher
+		const definition = getRegistryIdentifier(registry)
+		if (!definition) {
+			return
+		}
+		// Create a linked list representation
+		const leaf = {
+			type: 'leaf',
+			range: core.Range.create(node.range.end),
+		} as const
+		let link: NbtPathLink = { path: node, node: leaf }
+		for (let i = node.children.length - 1; i >= 0; i -= 1) {
+			link = {
+				path: node,
+				node: node.children[i],
+				next: link,
+			}
+		}
+		let prev = link
+		while (prev.next) {
+			prev.next.prev = prev
+			prev = prev.next
+		}
+		mcdoc.runtime.checker.reference<NbtPathLink>(
+			[{ originalNode: link, inferredType: inferPath(link) }],
+			definition,
+			{
+				context: ctx,
+				isEquivalent: (inferred, def) => {
+					switch (inferred.kind) {
+						case 'list':
+						case 'byte_array':
+						case 'int_array':
+						case 'long_array':
+							return [
+								'list',
+								'tuple',
+								'byte_array',
+								'int_array',
+								'long_array',
+							].includes(def.kind)
+						default:
+							return false
+					}
+				},
+				getChildren: (
+					link,
+				): mcdoc.runtime.checker.RuntimeUnion<NbtPathLink>[] => {
+					while (
+						link.next && link.node.type !== 'leaf' &&
+						NbtCompoundNode.is(link.node)
+					) {
+						link = link.next
+					}
+					if (!link.next || link.node.type === 'leaf') {
+						return []
+					}
+					if (NbtPathIndexNode.is(link.node)) {
+						return [[{
+							originalNode: link.next,
+							inferredType: inferPath(link.next),
+						}]]
+					}
+					if (core.StringNode.is(link.node)) {
+						return [{
+							key: {
+								originalNode: link,
+								inferredType: {
+									kind: 'literal',
+									value: { kind: 'string', value: link.node.value },
+								},
+							},
+							possibleValues: [{
+								originalNode: link.next,
+								inferredType: inferPath(link.next),
+							}],
+						}]
+					}
+					// Never reachable
+					return []
+				},
+				reportError: (error) => {
+					if (
+						error.kind === 'missing_key' ||
+						error.kind === 'invalid_collection_length'
+					) {
+						return
```

Comment:
why is this an empty return?

Reply:
Due to the way nbt paths are represented to the mcdoc checker (see `NbtPathLink`), the checker will report lots of "missing key" errors, because you're only visiting one branch of the children. Likewise when you're indexing into a list, there is no way for you to specify all N children.

Reply:
Ideally for `some_list[5]` you would create 4 `unsafe` types as children (or empty union?) and report a custom error when the maximum is exceeded.